File: backtracking/06_Combination_Sum/go/06_Combination_Sum.go
package main

import (
	"fmt"
	"reflect"
	"sort"
)

func main() {
	testCases := []struct {
		candidates []int
		target     int
		expect     [][]int
	}{
		{
			candidates: []int{2, 3, 6, 7},
			target:     7,
			expect:     [][]int{{2, 2, 3}, {7}},
		},
		{
			candidates: []int{2, 3, 5},
			target:     8,
			expect:     [][]int{{2, 2, 2, 2}, {2, 3, 3}, {3, 5}},
		},
		{
			candidates: []int{2},
			target:     1,
			expect:     [][]int{},
		},
		{
			candidates: []int{1},
			target:     1,
			expect:     [][]int{{1}},
		},
		{
			candidates: []int{1, 2},
			target:     4,
			expect:     [][]int{{1, 1, 1, 1}, {1, 1, 2}, {2, 2}},
		},
	}

	for i, testCase := range testCases {
		fmt.Printf("Test Case %d:\n", i+1)
		fmt.Printf("Input: candidates = %v, target = %d\n", testCase.candidates, testCase.target)

		result := combinationSum(testCase.candidates, testCase.target)

		fmt.Printf("Result: %v\n", result)
		fmt.Printf("Expect: %v\n", testCase.expect)

		// Sort inner slices and outer slice for comparison
		for j := range result {
			sort.Ints(result[j])
		}
		for j := range testCase.expect {
			sort.Ints(testCase.expect[j])
		}

		sort.Slice(result, func(i, j int) bool {
			return reflect.DeepEqual(result[i], result[j])
		})
		sort.Slice(testCase.expect, func(i, j int) bool {
			return reflect.DeepEqual(testCase.expect[i], testCase.expect[j])
		})

		if reflect.DeepEqual(result, testCase.expect) {
			fmt.Println("PASS")
		} else {
			fmt.Println("FAIL")
		}

		fmt.Println()
	}
}

func combinationSum(candidates []int, target int) [][]int {
	result := [][]int{}

	var combinationSumRecursive func(start int, current []int, currentSum int)
	combinationSumRecursive = func(start int, current []int, currentSum int) {
		// Base cases
		if currentSum > target {
			return
		}

		if currentSum == target {
			currentCopy := make([]int, len(current))
			copy(currentCopy, current)
			result = append(result, currentCopy)
			return
		}

		// Recursive case
		for i := start; i < len(candidates); i++ {
			combinationSumRecursive(i, append(current, candidates[i]), currentSum+candidates[i])
		}
	}

	combinationSumRecursive(0, []int{}, 0)
	return result
}

File: backtracking/06_Combination_Sum/06_Combination_Sum.js
const testCases = [
  {
    candidates: [2, 3, 6, 7],
    target: 7,
    expected: [[2, 2, 3], [7]],
  },
  {
    candidates: [2, 3, 5],
    target: 8,
    expected: [
      [2, 2, 2, 2],
      [2, 3, 3],
      [3, 5],
    ],
  },
  {
    candidates: [2],
    target: 1,
    expected: [],
  },
  {
    candidates: [1],
    target: 1,
    expected: [[1]],
  },
  {
    candidates: [1, 2],
    target: 4,
    expected: [
      [1, 1, 1, 1],
      [1, 1, 2],
      [2, 2],
    ],
  },
];

(function () {
  testCases.forEach((testCase, i) => {
    console.log(`Test Case ${i + 1}:`);
    console.log(
      `Input: candidates = ${JSON.stringify(testCase.candidates)}, target = ${
        testCase.target
      }`
    );
    const result = combinationSum(testCase.candidates, testCase.target);

    // Sort inner arrays for comparison
    const sortedResult = result
      .map((inner) => [...inner].sort((a, b) => a - b))
      .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
    const sortedExpected = testCase.expected
      .map((inner) => [...inner].sort((a, b) => a - b))
      .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));

    console.log(`Result: ${JSON.stringify(sortedResult)}`);
    console.log(`Expected: ${JSON.stringify(sortedExpected)}`);

    if (JSON.stringify(sortedResult) === JSON.stringify(sortedExpected)) {
      console.log(`Test case ${i + 1} passed`);
    } else {
      console.log(`Test case ${i + 1} failed`);
    }
    console.log();
  });
})();

function combinationSum(candidates, target) {
  let result = [];

  function combinationSumRecursive(start, current, currentSum) {
    // Base cases
    if (currentSum > target) return;

    if (currentSum == target) {
      result.push([...current]);
      return;
    }

    // Recursive case
    for (let i = start; i < candidates.length; i++) {
      current.push[candidates[i]];
      combinationSumRecursive(i, current, currentSum + candidates[i]);
      current.pop();
    }
  }

  combinationSumRecursive(0, [], 0);
  return result;
}
